qutebrowser/misc/overlap.py

```diff
@@ -0,0 +1,127 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2021 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Work with overlapping widgets."""
+
+import collections
+import math
+from typing import Dict, Iterable, List, Sequence, Set, Tuple
+
+from PyQt5.QtCore import QPoint, QRect
+from PyQt5.QtWidgets import QWidget
+
+
+def _corners(rect: QRect) -> List[QPoint]:
+    return [
+        rect.topLeft(), rect.topRight(), rect.bottomLeft(), rect.bottomRight()
+    ]
+
+
+class OverlapLookup:
+    """Hold a collection of widgets and compute overlaps.
+
+    Suppose we tile the viewport with cells of shape max_width x max_height.
+    If a widget intersects a cell, at least one of its corners are in the cell.
+    Then `self._lookup` maps (i,j) to set of widgets that intersect cell (i,j).
+    If widget1 intersects widget2 then widget2 must intersect at least one
+    cell that widget1 intersects; this allows us to avoid checking every pair
+    of widgets to see if they intersect.
+
+    alternatives: R-tree, quadtree
+    """
+
+    def __init__(self, widgets: Iterable[QWidget]):
+        self._lookup: Dict[
+            Tuple[int, int], Set[QWidget]
+        ] = collections.defaultdict(set)
+        self._max_width = max(widget.width() for widget in widgets)
+        self._max_height = max(widget.height() for widget in widgets)
+        for widget in widgets:
+            self.register(widget)
+
+    def _cell_index(self, p: QPoint) -> Tuple[int, int]:
+        return (
+            math.floor(p.x() / self._max_width),
+            math.floor(p.y() / self._max_height),
+        )
+
+    def register(self, widget: QWidget) -> None:
+        """Add a widget."""
+        for corner in _corners(widget.geometry()):
+            self._lookup[self._cell_index(corner)].add(widget)
+
+    def deregister(self, widget: QWidget) -> None:
+        """Remove a widget."""
+        for corner in _corners(widget.geometry()):
+            self._lookup[self._cell_index(corner)].difference_update([widget])
```

Comment:
Wouldn't

```suggestion
            self._lookup[self._cell_index(corner)].remove(widget)
```

do the same?